staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go

```diff
@@ -0,0 +1,550 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package immutable
+
+import (
+	"fmt"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	xListTypeMap    = "map"
+	xListTypeAtomic = "atomic"
+	xListTypeSet    = "set"
+	xMapType        = "granular"
+
+	genericTypeInteger      = "integer"
+	genericTypeNumber       = "number"
+	genericTypeString       = "string"
+	genericTypeBool         = "boolean"
+	genericTypeObject       = "object"
+	genericTypeArray        = "array"
+	xImmutabilityImmutable  = "immutable"
+	xImmutabilityAddOnly    = "addOnly"
+	xImmutabilityRemoveOnly = "removeOnly"
+)
+
+// Immutable validates immutable values are respected or error otherwise
+func Immutable(old interface{}, new interface{}, s *structuralschema.Structural, pth *field.Path) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+	switch old := old.(type) {
+	case map[string]interface{}:
+		newObj, ok := new.(map[string]interface{})
+		if !ok {
+			// e.g the CRD definition has changed
+			// TODO: what's the expectation here?
+		}
+		for k, prop := range s.Properties {
+			// XImmutable and XImmutableKeys are mutually exclusive
+			// CRD validation should never let this be true
+			if prop.XImmutability != nil && prop.XKeyImmutability != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "XImmutability and XKeyImmutability are mutually exclusive"))
+				continue
+			}
+
+			_, foundOld := old[k]
+			_, foundNew := newObj[k]
+
+			// XImmutability
+			// validate immutable / addOnly /removeOnly against the current property regardless of the type.
+			// XImmutability/XKeyImmutability below this level is disallowed.
+			if prop.XImmutability != nil {
+				if *prop.XImmutability == xImmutabilityImmutable {
+					if (foundOld != foundNew) || !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
+						allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					}
+					continue
+				}
+				if *prop.XImmutability == xImmutabilityAddOnly {
+					if foundOld && !(equality.Semantic.DeepEqual(old[k], newObj[k])) {
```

Comment:
what about: `foundOld==true`, `foundNew==false`, `old[k]=n=il` (i.e.  `null` in JSON), `newObj[k]==nil`). Then this does not trigger, but should.